### PR TITLE
rustdoc: give proc-macros their own pages

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
related to https://github.com/rust-lang/rust/issues/49553 but i don't think it'll fix it

Currently, rustdoc doesn't expose proc-macros all that well. In the source crate, only their definition function is exposed, but when re-exported, they're treated as a macro! This is an awkward situation in all accounts. This PR checks functions to see whether they have any of `#[proc_macro]`, `#[proc_macro_attribute]`, or `#[proc_macro_derive]`, and exposes them as macros instead. In addition, attributes and derives are exposed differently than other macros, getting their own item-type, CSS class, and module heading.

image

Function-like proc-macros are lumped in with `macro_rules!` macros, but they get a different declaration block (i'm open to tweaking this, it's just what i thought of given how function-proc-macros operate):

image

Proc-macro attributes and derives get their own pages, with a representative declaration block. Derive macros also show off their helper attributes:

image

image

There's one wrinkle which this PR doesn't address, which is why i didn't mark this as fixing the linked issue. Currently, proc-macros don't expose their attributes or source span across crates, so while rustdoc knows they exist, that's about all the information it gets. This leads to an "inlined" macro that has absolutely no docs on it, and no `[src]` link to show you where it was declared.

The way i got around it was to keep proc-macro re-export disabled, since we do get enough information across crates to properly link to the source page:

image

Until we can get a proc-macro's docs (and ideally also its source span) across crates, i believe this is the best way forward.

### CODE REVIEW

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -4598,6 +4608,17 @@ fn item_macro(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
     document(w, cx, it)
 }
 
+fn item_proc_macro(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item, m: &clean::ProcMacro)
+    -> fmt::Result
+{
+    if m.kind == MacroKind::Bang {
+        write!(w, "<pre class='rust macro'>")?;
+        write!(w, "{}!() {{ /* proc-macro */ }}", it.name.as_ref().unwrap())?;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render.rs]: No `unwrap` please (`expect`!).

**Comment by GuillaumeGomez** [comment]
That's quite the nice add, thanks a lot!

However, the UI looks a bit... empty? At least it tells readers that it exists but not much more...

**Comment by Centril** [comment]
> Function-like proc-macros are lumped in with macro_rules! macros

This is great; don't change it :)
I think it makes perfect sense for all bang macros to be lumped together.

**Comment by QuietMisdreavus** [comment]
> However, the UI looks a bit... empty? At least it tells readers that it exists but not much more...

@GuillaumeGomez Yeah, there's not much to say generically about proc-macros in general, other than their name and form factor. (I mean, it's technically possible for a macro author to create some kind of BASIC dialect in one of these proc-macros...) It's up to the author to fill in how it's supposed to be used. I can change up the attribute and derive pages to give them a "declaration" block, though. That should help guide those pages some. (It's possible to gather a custom derive's helper attribute, so i should definitely figure out how to display that one...)

> I think it makes perfect sense for all bang macros to be lumped together.

@Centril Yeah, at some point i realized there wasn't enough reason to separate `macro_rules!` macros and bang-proc-macros, so they're put under the same heading.

**Comment by QuietMisdreavus** [comment]
I've pushed up a couple new commits that add more information to attribute and derive pages. New images were added to the PR description to show off how they look now.

**Comment by GuillaumeGomez** [comment]
Seems good to me. A bit weird but as long as it's working... Please also add the proc-macros to the "all items" page (look for `AllTypes`).

**Comment by QuietMisdreavus** [comment]
Good catch! I've added sections to the "All items" page for attributes and derives, since function proc-macros already appear in the existing "Macros" section. I also found another issue where the link to the "all items" page was only being printed if the `--crate-version` argument was sent, so i went ahead and fixed that here.

**Comment by GuillaumeGomez** [comment]
Thanks for the fix. Please also add a test to check if they are correctly listed in the all items page (just add it to the all items test). Once done and CI green, r=me.

**Comment by QuietMisdreavus** [comment]
I added it to the proc-macro links test, so that i wouldn't have to mess with the crate type of the all-items test.

@bors r=GuillaumeGomez

**Pull request merged**